internal/xds/rbac/rbac_engine_test.go

```
@@ -1007,85 +1224,704 @@ func (s) TestChainEngine(t *testing.T) {
+type TestAuditLoggerBufferConfig struct {
+	audit.LoggerConfig
+}
+
+func (builder *TestAuditLoggerBufferBuilder) ParseLoggerConfig(configJSON json.RawMessage) (config audit.LoggerConfig, err error) {
+	return TestAuditLoggerBufferConfig{}, nil
+}
+
+func (builder *TestAuditLoggerBufferBuilder) Build(config audit.LoggerConfig) audit.Logger {
+	return &TestAuditLoggerBuffer{auditEvents: &builder.auditEvents}
+}
+
+func (builder *TestAuditLoggerBufferBuilder) Name() string {
+	return "TestAuditLoggerBuffer"
+}
+
+// An audit logger to test using a custom config
+type TestAuditLoggerCustomConfig struct{}
+
+func (logger *TestAuditLoggerCustomConfig) Log(*audit.Event) {}
+
+// Build TestAuditLoggerCustomConfig
+type TestAuditLoggerCustomConfigBuilder struct {
+}
+
+// The custom config for the TestAuditLoggerCustomConfig logger
+type TestAuditLoggerCustomConfigConfig struct {
+	audit.LoggerConfig
+	Abc int
+	Xyz string
+}
+
+// Parses TestAuditLoggerCustomConfigConfig
+// Hard-coded to match with it's test case above
+func (builder TestAuditLoggerCustomConfigBuilder) ParseLoggerConfig(configJSON json.RawMessage) (audit.LoggerConfig, error) {
+	c := TestAuditLoggerCustomConfigConfig{}
+	err := json.Unmarshal(configJSON, &c)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse custom config: %v", err)
+	}
+	return c, nil
+}
+
+func (builder *TestAuditLoggerCustomConfigBuilder) Build(config audit.LoggerConfig) audit.Logger {
+	return &TestAuditLoggerCustomConfig{}
+}
+
+func (builder *TestAuditLoggerCustomConfigBuilder) Name() string {
+	return "TestAuditLoggerCustomConfig"
+}
+
+// Builds custom configs for audit logger RBAC protos
+func createUDPATypedStruct(t *testing.T, in map[string]interface{}, name string) *anypb.Any {
+	t.Helper()
+	pb, err := structpb.NewStruct(in)
+	typedURL := ""
+	if name != "" {
+		typedURL = typedURLPrefix + name
+	}
+	typedStruct := &v1xdsudpatypepb.TypedStruct{
+		TypeUrl: typedURL,
+		Value:   pb,
+	}
+	if err != nil {
```

Its better to fail fast by moving this check right after the `NewStruct()` call above

---

Good catch